### PR TITLE
[HttpFoundation] [PSR-7] Allow to use resources as content body and to return resources from string content

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | not yet |
- Allows to fix tests of https://github.com/symfony/psr-http-message-bridge with PHP 5.6.
- Ease the transition to PSR-7 (in PSR-7, almost everything is stream - #15186) 

Maybe should I open it against 2.8 but it can be considered a bug fix at least for the part "returning a string as a resource".

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -1465,16 +1465,36 @@ public function isMethodSafe()
...
     public function getContent($asResource = false)
     {
-        if (PHP_VERSION_ID < 50600 && (false === $this->content || (true === $asResource && null !== $this->content))) {
+        $currentContentIsResource = is_resource($this->content);
+        if (PHP_VERSION_ID < 50600 && (false === $this->content || (true === $asResource && null !== $this->content && !$currentContentIsResource))) {
             throw new \LogicException('getContent() can only be called once when using the resource return type and PHP below 5.6.');
         }
 
         if (true === $asResource) {
+            if (is_resource($this->content)) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Request.php]: you can reuse `$currentContentIsResource`

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -1465,16 +1465,36 @@ public function isMethodSafe()
...
+
+                return $resource;
+            }
+
             $this->content = false;
 
             return fopen('php://input', 'rb');
         }
 
+        if (is_resource($this->content)) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Request.php]: same here

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 3]
```diff
@@ -1465,16 +1465,36 @@ public function isMethodSafe()
      */
     public function getContent($asResource = false)
     {
-        if (PHP_VERSION_ID < 50600 && (false === $this->content || (true === $asResource && null !== $this->content))) {
+        $currentContentIsResource = is_resource($this->content);
+        if (PHP_VERSION_ID < 50600 && (false === $this->content || (true === $asResource && null !== $this->content && !$currentContentIsResource))) {
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Request.php]: Is the check for `$currentContentIsResource` really necessary? I mean is PHP not capable to handle two streams reading from `php://temp` (I would have guessed that the temporary filenames would differ when using it multiple times)?

**Comment by dunglas** [comment]
Should be OK now.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 4]
```diff
@@ -1465,16 +1465,38 @@ public function isMethodSafe()
...
         if (true === $asResource) {
+            if ($currentContentIsResource) {
+                rewind($this->content);
+
+                return $this->content;
+            }
+
+            // Content passed in parameter (test)
+            if (is_string($this->content)) {
+                $resource = fopen('php://temp','r+');
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/Request.php]: missing space after the comma.

**Comment by fabpot** [comment]
:+1:

**Comment by stof** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @dunglas.

**Pull request closed without merging**